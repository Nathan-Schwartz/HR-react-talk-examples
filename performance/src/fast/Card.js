import React from 'react';
import {Card, CardActions, CardHeader, CardTitle, CardText} from 'material-ui/Card';
import FlatButton from 'material-ui/FlatButton';
import Toggle from 'material-ui/Toggle';

export default class ContactCard extends React.Component {

  state = {
    expanded: false,
  }

  // Here we are using shouldComponentUpdate to make sure that
  // we only rerender when values change
  shouldComponentUpdate(nextProps, nextState) {
    return nextProps.email !== this.props.email
      || nextState.expanded !== this.state.expanded
  }

  handleToggle = (event, toggle) => {
    this.setState({expanded: toggle});
  };

  noopClick = () => {}

  render() {
    console.log("Card Render")
    const { email, first_name, last_name, image } = this.props;
    return (
      <span style={{ width: 250, display: 'inline-block', margin: 10 }}>
        <Card>
          <CardHeader
            title={`${first_name} ${last_name}`}
            subtitle={email}
            avatar={image}
          />
          <CardText>
            <Toggle
              toggled={this.state.expanded}
              onToggle={this.handleToggle}
              style={{ marginLeft: 20 }}
              labelPosition="right"
              label="random toggle"
            />
          </CardText>
          <CardActions>
            <FlatButton label="Expand" onTouchTap={this.noopClick} />
            <FlatButton label="Reduce" onTouchTap={this.noopClick} />
          </CardActions>
        </Card>
      </span>
    );
  }
}