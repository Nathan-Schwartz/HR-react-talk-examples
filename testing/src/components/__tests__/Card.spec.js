import React from 'react';
import { mount } from 'enzyme';
import { spy } from 'sinon';
import toJSON from 'enzyme-to-json';

import Card from '../Card';


const mockProps = {
  first_name: 'John',
  last_name: 'Smith',
  email: 'john.smith@email.com',
  image: 'https://www.notarealurl.com/',
  handleImportantClick: spy(),
}

describe('', () => {
  let mountedCard;
  it('should mount', () => {
    // Render and return an object representation of the 
    // entire Card subtree.
    mountedCard = mount(
      <Card {...mockProps} />,
    );

    // Should mount.
    expect(mountedCard).toBeTruthy();
  });

  it('Should match snapshot', () => {
    expect(toJSON(mountedCard)).toMatchSnapshot();

    // This eliminates the following manual tests:
    // Should recieve X props
    // Should render a span, with a div inside, and two buttons in that.
    // Should display X text in each button.
    // Should use X styles.

    // // Some of those tests would look like this:

    // // Outermost element isn't considered a "child"
    // const Span = mountedCard.find('span');
    // expect(Span.length).toBe(1);

    // expect(mountedCard.childAt(0).type()).toBe('div')
    // expect(mountedCard.childAt(0).children().map(a => a.type()))
    //   .toEqual(["img", "button","button"])

    // const Buttons = mountedCard.find('button')
    // expect(Buttons.length).toBe(2);
    // expect(Buttons.map(button => button.text()))
    //   .toEqual(["Important Button!", "Counter Button!"])
  });

  it('Should call handleImportantClick prop', () => {
    expect(mockProps.handleImportantClick.callCount).toBe(0);
    
    const ImportantButton = mountedCard
      .findWhere(a => a.props().onClick === mountedCard.props().handleImportantClick);

    expect(ImportantButton.length).toBe(1);
    ImportantButton.simulate('click');

    expect(mockProps.handleImportantClick.callCount).toBe(1);
  })
  
  it('Increment method should increment state', () => {
    expect(mountedCard.state().counter).toBe(0);
    mountedCard.instance().handleIncrement();
    expect(mountedCard.state().counter).toBe(1);
  })

  it('Clicking CounterButton should call handleIncrement', () => {

    // Inject spy.
    spy(mountedCard.instance(), "handleIncrement");

    // Update the component so that the component's uses of handleIncrement
    // don't still refer to the original.
    mountedCard.update();
    
    expect(mountedCard.instance().handleIncrement.callCount).toBe(0);

    // Retrieve the button that uses it
    const CounterButton = mountedCard
      .findWhere(a => a.props().onClick === mountedCard.instance().handleIncrement);

    expect(CounterButton.length).toBe(1);

    // Click on it
    CounterButton.simulate('click')

    // Make sure the spy was called
    expect(mountedCard.instance().handleIncrement.callCount).toBe(1);
  })
});