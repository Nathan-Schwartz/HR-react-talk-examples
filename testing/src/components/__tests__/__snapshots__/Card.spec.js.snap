// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should match snapshot 1`] = `
<ContactCard
  email="john.smith@email.com"
  first_name="John"
  handleImportantClick={[Function]}
  image="https://www.notarealurl.com/"
  last_name="Smith"
>
  <span
    style={
      Object {
        "display": "inline-block",
        "margin": 10,
        "width": 250,
      }
    }
  >
    <div>
      <img
        src="https://www.notarealurl.com/"
      />
      <button
        onClick={[Function]}
      >
        Important Button!
      </button>
      <button
        onClick={[Function]}
      >
        Counter Button!
      </button>
    </div>
  </span>
</ContactCard>
`;
